### PR TITLE
Add GRPC function to get points in box

### PR DESCRIPTION
**Pull request by thomasschiwietz**

### CODE REVIEW

File: grpc/src/bin/octree_server.rs [hunk 1]
```diff
@@ -110,6 +113,40 @@ impl proto_grpc::Octree for OctreeService {
...
+        let mut replies = Vec::new();
+        replies.push((proto::GetPointsInBoxReply::new(), WriteFlags::default()));
+        self.octree.points_in_box(&bounding_box).for_each(|p| {
+            let mut v = point_viewer::proto::Vector3f::new();
+            v.set_x(p.position.x);
+            v.set_y(p.position.y);
+            v.set_z(p.position.z);
+            replies.last_mut().unwrap().0.mut_points().push(v);
+            // message but be below 4 MB
+            if replies.last().unwrap().0.points.len() > 240000 {
```
Comment (SirVer) [inline grpc/src/bin/octree_server.rs]: Add an explanation how you decided on the number.

File: grpc/src/lib.rs [hunk 2]
```diff
@@ -55,6 +58,33 @@ impl GrpcOctree {
             root_bounding_cube,
         }
     }
+
+    pub fn get_points_in_box(&self, bounding_box: &Aabb3<f32>) -> Vec<Point3<f32>> {
```
Comment (SirVer) [inline grpc/src/lib.rs]: add a TODO that this should return errors

File: grpc/src/lib.rs [hunk 3]
```diff
@@ -55,6 +58,33 @@ impl GrpcOctree {
...
+        req.mut_bounding_box().mut_min().set_x(bounding_box.min.x);
+        req.mut_bounding_box().mut_min().set_y(bounding_box.min.y);
+        req.mut_bounding_box().mut_min().set_z(bounding_box.min.z);
+        req.mut_bounding_box().mut_max().set_x(bounding_box.max.x);
+        req.mut_bounding_box().mut_max().set_y(bounding_box.max.y);
+        req.mut_bounding_box().mut_max().set_z(bounding_box.max.z);
+        let mut replies = self.client.get_points_in_box(&req).unwrap();
+        let mut points = Vec::new();
+        loop {
+            let f = replies.into_future();
```
Comment (SirVer) [inline grpc/src/lib.rs]: use replies.for_each() which returns a future on which you can then wait. No need for the loop.

File: grpc/src/lib.rs [hunk 4]
```diff
@@ -124,3 +154,16 @@ impl Octree for GrpcOctree {
         Ok(result)
     }
 }
+
+#[cfg(test)]
```
Comment (SirVer) [inline grpc/src/lib.rs]: remove? I think we could start writing integration tests, i.e. bringing up an octree server with a on-disk octree and then sending rpc's back and forth. Open an issue for that?

Comment (thomasschiwietz) [reply grpc/src/lib.rs]: #103

File: grpc/src/bin/octree_server.rs [hunk 5]
```diff
@@ -110,6 +113,41 @@ impl proto_grpc::Octree for OctreeService {
...
+        let mut replies = Vec::new();
+        replies.push((proto::GetPointsInBoxReply::new(), WriteFlags::default()));
+        self.octree.points_in_box(&bounding_box).for_each(|p| {
+            let mut v = point_viewer::proto::Vector3f::new();
+            v.set_x(p.position.x);
+            v.set_y(p.position.y);
+            v.set_z(p.position.z);
+            replies.last_mut().unwrap().0.mut_points().push(v);
+            // Proto message must be below 4 MB.
+            // 240000 points take about 2.9 MB and proto requires about 1 MB of overhead for that.
```
Comment (SirVer) [inline grpc/src/bin/octree_server.rs]: If we ever change the points proto, this heuristic will fail. Can we get this more principled somehow?

File: grpc/src/lib.rs (hunk 4 above)
Comment (SirVer) [reply grpc/src/lib.rs]: please remove commented out code too.

**Comment by SirVer** [comment]
I can't think of a better solution right now....

**Comment by SirVer** [comment]
@wally-the-cartographer merge

**Comment by thomasschiwietz** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user thomasschiwietz. Merge queue now has a length of 1.

**Pull request merged**